File: test/scripts/setup-storage.test.ts
import { afterEach, describe, expect, test } from "bun:test";
import { mkdtempSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { setupStorage } from "../../scripts/tasks/setup-storage";

const touched: string[] = [];

afterEach(() => {
  for (const dir of touched.splice(0, touched.length)) {
    rmSync(dir, { force: true, recursive: true });
  }
  delete process.env.SCRY_INFRA_ENV_EXAMPLE;
  delete process.env.SCRY_INFRA_ENV_FILE;
});

describe("setupStorage", () => {
  test("adds missing keys from env example", () => {
    const dir = mkdtempSync(join(tmpdir(), "scry-storage-test-"));
    touched.push(dir);

    const envExample = join(dir, ".env.example");
    const envFile = join(dir, ".env");

    writeFileSync(envExample, "A=1\nB=2\nC=3\n");
    writeFileSync(envFile, "A=7\n\n# keep this comment\n");

    process.env.SCRY_INFRA_ENV_EXAMPLE = envExample;
    process.env.SCRY_INFRA_ENV_FILE = envFile;

    setupStorage();

    const content = readFileSync(envFile, "utf8");
    expect(content).toContain("A=7\n");
    expect(content).toContain("B=2\n");
    expect(content).toContain("C=3\n");
  });
});

File: test/setup_storage_test.rb
# frozen_string_literal: true

require 'fileutils'
require 'tmpdir'
require_relative 'test_helper'

class SetupStorageTaskTest < Minitest::Test
  include Scry::TestHelpers

  def test_adds_missing_keys_from_env_example
    Dir.mktmpdir('scry-storage-test-') do |tmp|
      env_example = File.join(tmp, '.env.example')
      env_file = File.join(tmp, '.env')

      File.write(env_example, "A=1\nB=2\nC=3\n")
      File.write(env_file, "A=7\n\n# keep this comment\n")

      with_env(
        'SCRY_INFRA_ENV_EXAMPLE' => env_example,
        'SCRY_INFRA_ENV_FILE' => env_file
      ) do
        task = Rake::Task['scry:setup:storage']
        task.reenable
        task.invoke
      end

      content = File.read(env_file)

      assert_includes content, "A=7\n"
      assert_includes content, "B=2\n"
      assert_includes content, "C=3\n"
    end
  end
end
